pkg/beacon/relay/groupselection/groupselection.go

```diff
@@ -87,192 +92,188 @@ func CandidateToNewGroup(
 }
 
 func startTicketSubmission(
-	initialSubmissionTickets []*ticket,
-	reactiveSubmissionTickets []*ticket,
+	tickets []*ticket,
 	relayChain relaychain.GroupSelectionInterface,
 	blockCounter chain.BlockCounter,
 	chainConfig *config.Chain,
 	startBlockHeight uint64,
 	onGroupSelected func(*Result),
```

Comment:
Since this function blocks until ticket submission is over, we don't need to pass `onGroupSelected` as a parameter and execute this function in `CandidateToNewGroup` instead. I think it's much cleaner approach:
```
func CandidateToNewGroup(
  // (...)
) error {
  tickets, err := generateTickets(...)
  // (...)
  err = submitTickets(...)
  // (...)
  
  selectedStakers, err := relayChain.GetSelectedParticipants()
  if err != nil {
    return fmt.Errorf(
      "could not fetch selected participants after submission timeout [%v]",
      err,
    )
  }

  go onGroupSelected(&Result{
    SelectedStakers:        selectedStakers,
    GroupSelectionEndBlock: ticketSubmissionEndBlockHeight,
  })
}
 
```

Reply:
It makes even more sense if we consider how we should handle errors from `err = submitTickets(...)`. I don't think we should `return` but just `logger.Errorf` and continue with the execution of `CandidateToNewGroup`. See that in case of e.g. `WaitForBlockHeight` in round 4 of ticket submission, there is a chance tickets from rounds 1-3 qualified to a group. It does make sense to stop `submitTickets` processing on any error but we should still wait for ticket submission timeout in `CandidateToNewGroup` and execute `GetSelectedParticipants`:

```
err = submitTickets(...)
if err != nil {
  logger.Errorf(...). // <--- !!!
}

// wait till the end of the ticket submission in case submitTickets failed in the middle
// and there is still a chance we qualified to a group
ticketSubmissionTimeoutChannel, err := blockCounter.BlockHeightWaiter(
  startBlockHeight + chainConfig.TicketSubmissionTimeout,
)
if err != nil {
  return err
}
  
selectedStakers, err := relayChain.GetSelectedParticipants()
if err != nil {
    return fmt.Errorf(
      "could not fetch selected participants after submission timeout [%v]",
      err,
    )
}
```

What do you think about it?

Reply:
Excellent idea! Will change as you proposed.